src/attacks.rs

```diff
@@ -14,16 +20,171 @@ pub enum DepthReduceSet {
+                AttackTarget::Size(size) => size,
+            };
+            // FIXME: Too verbose, there probably is a more concise way to do this.
+            TargetRange {
+                start: single_value,
+                end: single_value,
+                interval: 0.0,
+            }
+        };
+
+        AttackProfile {
+            runs: 1,
+            target,
+            range,
+            attack,
+        }
+    }
+}
+
+/// Results of an attack expressed in relation to the graph size.
+#[derive(Clone, Default)]
+pub struct SingleAttackResult {
+    depth: f64,
+    exclusion_size: f64,
+    // graph_size: usize,
+    // FIXME: Do we care to know the absolute number or just
+    // relative to the graph size?
+}
+
+impl std::fmt::Display for SingleAttackResult {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "\t-> |S| = {:.4}\n\t-> depth(G-S) = {:.4}",
+            self.exclusion_size, self.depth,
+        )
+    }
+}
+
+pub fn attack(g: &mut Graph, attack: DepthReduceSet) -> SingleAttackResult {
+    let start = Instant::now();
+    let set = depth_reduce(g, attack);
+    let duration = start.elapsed();
+    let depth = g.depth_exclude(&set);
+    let result = SingleAttackResult {
+        depth: depth as f64 / g.size() as f64,
+        exclusion_size: set.len() as f64 / g.size() as f64,
+    };
+    println!("{}", result);
+    println!("\t-> time elapsed: {:?}", duration);
+    result
+}
+
+// FIXME: Eventually this should replace the old `attack`.
+pub fn attack_with_profile(g: &mut Graph, profile: AttackProfile) {
+    let mut targets: Vec<f64> = Vec::new();
+    let mut target = profile.range.start;
+    loop {
+        targets.push(target);
+        target += profile.range.interval;
+        if target >= profile.range.end {
+            break;
+        }
     }
+    // FIXME: Move this logic to `TargetRange`.
+
+    let mut results: Vec<Vec<SingleAttackResult>> =
+        vec![vec![SingleAttackResult::default(); profile.runs]; targets.len()];
+
+    for (t, target) in targets.iter().enumerate() {
+        let absolute_target = (target * g.size() as f64) as usize;
+        let attack_type = match profile.attack.clone() {
+            DepthReduceSet::ValiantDepth(_) => DepthReduceSet::ValiantDepth(absolute_target),
+            DepthReduceSet::ValiantSize(_) => DepthReduceSet::ValiantSize(absolute_target),
+            DepthReduceSet::ValiantAB16(_) => DepthReduceSet::ValiantAB16(absolute_target),
+            DepthReduceSet::GreedyDepth(_, p) => DepthReduceSet::GreedyDepth(absolute_target, p),
+            DepthReduceSet::GreedySize(_, p) => DepthReduceSet::GreedySize(absolute_target, p),
+        };
+        // FIXME: Same as before, the target should be decoupled from the type of attack.
+
+        for run in 0..profile.runs {
+            // FIXME: Should we create a new graph for different runs? Or targets?
```

Comment:
^

Reply:
We should generated `profile.runs` graphs, and for each of these graphs, we should run the attacks for the different depth/set size. Does that answer the comment ?

Reply:
This seems to be a bit contradictory with https://github.com/filecoin-project/drg-attacks/pull/23#discussion_r332973439

> output all data

So to refine the objective a bit, we average across `runs` _inside_  (collapsing that dimension) and output an array of a single dimensions with different targets, does that sound right?

Reply:
> We should generate `profile.runs` graphs, and for each of these graphs, we should run the attacks for the different depth/set size.

Should then the `size`, `seed` and `DRGAlgo` be an input to the attack (probably encapsulated in a `GraphSpec` structure) instead of a single generated `Graph` (as is now), included in the profile, and from there the multiple graphs generated to run the attack on?

If that sounds right, one should save the graphs _with_ the `GraphSpec` data (to be able to identify those options for future runs, probably with a hash).